sycl/include/sycl/ext/intel/experimental/esimd/kernel_properties.hpp

```diff
@@ -0,0 +1,78 @@
+//==---------------- common.hpp - DPC++ Explicit SIMD API   ----------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// APIs for setting kernel properties interpreted by GPU software stack.
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/ext/intel/experimental/esimd/common.hpp>
+#include <sycl/ext/intel/experimental/esimd/detail/misc_intrin.hpp>
+
+#include <sycl/detail/boost/mp11.hpp>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace __ESIMD_ENS {
+
+namespace kernel_properties {
+
+/// <summary>
+///  This namespace contains APIs to set kernel properties.
+/// </summary>
+
+// Implementation note: <property_class>::value fields should match property IDs
+// specified in llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp
+
+namespace detail {
+// Proxy to access private property classes' fields from the API code.
+template <class T> struct proxy {
+  static constexpr int value = T::value;
+};
+} // namespace detail
+
+/// A boolean property which requests the compiler to double the amount of
+/// general-purpose registers available to a thread at the expense of reducing
+/// the amount of available hardware threads.
+struct use_double_grf_tag {
+  template <class T> friend struct detail::proxy;
+
+private:
+  // Property identifier
+  static constexpr int value = 0;
+};
+
+inline constexpr use_double_grf_tag use_double_grf = {};
+
+} // namespace kernel_properties
+
+namespace __MP11_NS = sycl::detail::boost::mp11;
+
+// TODO this should be replaced with the generic SYCL compile-time properites
+// mechanism once implementaion is available.
+// https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_kernel_properties.asciidoc
+
+template <class... KernelProps>
+void set_kernel_properties(KernelProps... props) {
+  // TODO check for duplicates
+  using Props = __MP11_NS::mp_list<KernelProps...>;
```

Comment:
Is this the first time we use boost to implement something in ESIMD?
If Yes, then it introduces the dependency on it, which probably, is not good without a good reason. IMO, it would be better to re-write this code without boost to not introduce that dependency for such simple thing as walk through variadic pack.

Reply:
Boost was imported quite a while ago to facilitate `invoke_simd` implementation and avoid spending resources on what boost already does. This is believed to be major productivity enhancement for SYCL API development. So this is first time for ESIMD, but not the first time for SYCL, so no new dependence is introduced. Note that `boosl::mp11` appears as `sycl::detail::boost::mp11` in SYCL API sources (it is imported and refactored automatically during the build)

Reply:
+1 to what @v-klochkov  suggested. Pure C++ solution looks simple enough: https://godbolt.org/z/fxv4jxPbb and if doesn't satisfy what needs to be done, then a comment explaining what and why would be very desirable.

Reply:
I don't think pure C++ solution should be chosen over mp11-based one just because it does not use mp11. Code clarity and maintainability is more important here, and mp11 makes it better - `list` + `for_each` is the most clear possible, yet adding new features to the code like duplicate detection would be way easier with mp11 as well.

In general, I believe it is much more practical and safer to use mp11 for all routine tasks where SYCL headers conceptually use templates as usual data structures as in this case, even though in some cases simpler C++ is possible.

Adding @rolandschulz for possible opinion on mp11 usage direction.